### PR TITLE
Add JobFailureReason enum to job failure event

### PR DESCRIPTION
**Pull request by gibsondan**
Summary:
This helps us track the specific reason a job failed in a structured way, which will allow us to make more informed run retry decisions (for example, not retrying if it was only do to an 'expected' op failure, which might have already retried sufficiently).

Test Plan: BK

### CODE REVIEW

File: python_modules/dagster/dagster/_core/storage/runs/sql_run_storage.py [hunk 1]
```diff
@@ -204,6 +212,14 @@ def handle_run_event(self, run_id: str, event: DagsterEvent) -> None:
                 )
             )
 
+        if event.event_type == DagsterEventType.PIPELINE_FAILURE and isinstance(
+            event.event_specific_data, JobFailureData
+        ):
+            failure_reason = event.event_specific_data.failure_reason
+
+            if failure_reason and failure_reason != RunFailureReason.UNKNOWN:
+                self.add_run_tags(run_id, {RUN_FAILURE_REASON_TAG: failure_reason.value})
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/runs/sql_run_storage.py:221]: do we already call `add_run_tags` for every run on some path? just making sure this isnt going to be substantial new load.

**Review by alangenfeld — CHANGES_REQUESTED** [review]
for this run tag bit we should add a test to the shared run storage test suite to ensure all storages respect it

File: python_modules/dagster/dagster/_core/storage/runs/sql_run_storage.py (hunk 1 above)
Comment (gibsondan) [reply python_modules/dagster/dagster/_core/storage/runs/sql_run_storage.py:221]: I think its called a couple of different times during the launch flow already - once at submit time, once at enqueue time at least

Comment (gibsondan) [reply python_modules/dagster/dagster/_core/storage/runs/sql_run_storage.py:221]: if we're worried we could potentially gate it on the new run retries flag I am planning to add that ingests it

Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/storage/runs/sql_run_storage.py:221]: I couldn't remember if it was just in the local grpc case or if it happened in other contexts as well

i think you are right and this should be fine

**Review by alangenfeld — APPROVED** [review]
👍

**Pull request merged**